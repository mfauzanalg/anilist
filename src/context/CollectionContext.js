import { useState, createContext, useEffect } from 'react';

export const CollectionContext = createContext();
export const CollectionProvider = (props) => {

  const [collections, setCollections] = useState(JSON.parse(localStorage.getItem('collections')) || [])

  useEffect(() => {
    localStorage.setItem('collections', JSON.stringify(collections))
  }, [collections])

  const addNewCollection = (name) => {
    const temp = [...collections]
    temp.push({ name: name })

    setCollections(temp);
  }

  return (
    <CollectionContext.Provider value={{ collections, addNewCollection }}>
      {props.children}
    </CollectionContext.Provider>
  )
}