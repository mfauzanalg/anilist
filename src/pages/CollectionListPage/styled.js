import styled from '@emotion/styled'

export const Title = styled.div`
  padding-top: 20px;
  font-size: 40px;
  font-weight: bold;
`

export const ComponentContainer = styled.div`
  padding: 20px 30px;
  position: relative;
  background-color: #141414;
  min-height: 100vh;
  background-image: linear-gradient(#a31515, transparent, transparent);
  position: relative;
`

export const CollectionListContainer = styled.div`
  margin-top: 30px;
  border-radius: 10px;
  display: flex;
  justify-content: space-between;
  flex-wrap: wrap;
  margin-bottom: 150px;

  &:after{
    content: ' ';
    width: 32%;
  }

`

export const ButtonContainer = styled.div`
  display: flex;
  margin-top: 5px;
  justify-content: space-between;
`

export const InputContainer = styled.div`
  display: flex;
  justify-content: center;
  width: 100%;
  margin-bottom: 15px;
`
