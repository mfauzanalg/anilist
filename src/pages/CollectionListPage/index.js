import { CollectionListContainer, ComponentContainer, Title } from './styled'
import CollectionCard from '../../components/CollectionCard'
import { CollectionContext } from '../../context/CollectionContext'
import { useContext } from 'react'

const CollectionListPage = () => {
  const { collections, addNewCollection } = useContext(CollectionContext)

  const addCollection = () => {
    addNewCollection('Saya')
  }

  return (
    <ComponentContainer>
      <Title onClick={addCollection}>
        Collection List
      </Title>
      <CollectionListContainer>
        {
          collections && collections.map((col, index) => {
            return (
              <CollectionCard title={col.name} key={index} />
            )
          })
        }
      </CollectionListContainer>
    </ComponentContainer>
  )
}

export default CollectionListPage