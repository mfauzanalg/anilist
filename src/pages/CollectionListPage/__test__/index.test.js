import { render, screen } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect';
import { MemoryRouter } from 'react-router-dom'
import CollectionListPage from '..'
import { CollectionProvider } from '../../../context/CollectionContext'
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';

const component = (
  <CollectionProvider>
    <MemoryRouter>
      <CollectionListPage />
    </MemoryRouter>
  </CollectionProvider>
)

const renderComponent = () => {
  return render(component)
}

it('CollectionListPage render correctly', () => {
  renderComponent();
})

it('If collection is empty', () => {
  renderComponent();
  expect(screen.getByText('Empty List of Collection')).toBeInTheDocument();

  const tree = renderer.create(component).toJSON;
  expect(tree).toMatchSnapshot();
})


it('Integration test on ContainerPageList', () => {
  renderComponent();

  // Add item to list
  userEvent.click(screen.queryByTestId('floating-button'));

  // expect modal is appear
  expect(screen.queryByTestId('modal')).toHaveTextContent('Create New Collection');

  // click cancel
  userEvent.click(screen.getByText('Cancel'))

  // expect modal is dissappear
  expect(screen.queryByTestId('modal')).toBeNull();

  // Click add again
  userEvent.click(screen.queryByTestId('floating-button'));

  // fill with special character
  userEvent.type(screen.getByPlaceholderText('Collection Name'), 'special char!')

  // click create and should not create any collection
  userEvent.click(screen.getByText('Create'));
  expect(screen.getByText('Empty List of Collection')).toBeInTheDocument();

  // fill with valid name
  userEvent.clear(screen.getByPlaceholderText('Collection Name'))
  userEvent.type(screen.getByPlaceholderText('Collection Name'), 'AnimeCollection')

  // click create
  userEvent.click(screen.getByText('Create'));

  // collection should be created
  expect(screen.queryAllByTestId('collection-card')).toHaveLength(1);
  expect(screen.getByTestId('collection-card').getAttribute('value')).toBe('AnimeCollection')

  // modal should be closed after success
  expect(screen.queryByTestId('modal')).toBeNull();

  // create collection with same name
  userEvent.click(screen.queryByTestId('floating-button'));
  userEvent.type(screen.getByPlaceholderText('Collection Name'), 'AnimeCollection');
  userEvent.click(screen.getByText('Create'));

  // Collection should not be created
  expect(screen.queryAllByTestId('collection-card')).toHaveLength(1);
  userEvent.click(screen.getByText('Cancel'));

  // Edit the collection
  userEvent.click(screen.getByTestId('edit-button'));
  userEvent.clear(screen.getByPlaceholderText('Collection Name'))
  userEvent.type(screen.getByPlaceholderText('Collection Name'), 'AnimeCollectionNew')

  // Click edit
  userEvent.click(screen.getByText('Edit'));

  // The collection name should be changed
  expect(screen.queryAllByTestId('collection-card')).toHaveLength(1);
  expect(screen.getByTestId('collection-card').getAttribute('value')).toBe('AnimeCollectionNew')

  // Modal should be closed
  expect(screen.queryByTestId('modal')).toBeNull();

  // Delete collection
  userEvent.click(screen.getByTestId('delete-button'));
  userEvent.click(screen.getByText('Delete'));

  // Collection should be empty again
  expect(screen.queryAllByTestId('collection-card')).toHaveLength(0);
  expect(screen.getByText('Empty List of Collection')).toBeInTheDocument();

  // Modal should be closed
  expect(screen.queryByTestId('modal')).toBeNull();
})

