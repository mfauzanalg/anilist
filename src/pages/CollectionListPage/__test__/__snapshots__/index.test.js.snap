// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`If collection is empty 1`] = `[Function]`;
